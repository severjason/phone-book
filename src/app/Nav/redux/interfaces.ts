import { AppContact } from '../../Contact/interfaces';
import { AppHomeAction } from '../../Home/interfaces';

export interface AppNavState {
  inputValue: string;
}

export interface AppSearchProps {
  isLoading: boolean;
  contacts: AppContact[];
  searchInput?: string;
}

export interface AppNavProps {
  inputValue: string;
}

export interface AppNavAction {
  type: string;
  payload?: string;
}

export interface AppSearchActions {
  startSearch: (input: string) => AppNavAction;
  finishSearch: () => AppNavAction;
  fetchContacts: () => AppHomeAction;
}