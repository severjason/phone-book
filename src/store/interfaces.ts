import { AppHomeState } from '../app/Home/interfaces';
import { AppNavState } from '../app/Nav/redux/interfaces';

export interface AppState {
  contactsState: AppHomeState;
  searchState: AppNavState;
}
