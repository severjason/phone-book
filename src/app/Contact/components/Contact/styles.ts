import styled from 'styled-components';

const ContactStyles = styled.div`
  margin: .5rem 0;
  
  .card-action {
    width: 100%;
  }
  .phones-container {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-wrap: nowrap;
  }
  
  .contact-content {
    display: none;
    padding: 1.5rem;
    &.expanded {
      display: block;
    }
    &:last-child {
      padding-bottom: 1.5rem;
    }
  }
`;

export default ContactStyles;
