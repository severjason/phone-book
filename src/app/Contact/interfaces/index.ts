
export interface AppContact {
  id: number;
  name: {
    first: string;
    last: string;
  };
  phone: string[];
}
