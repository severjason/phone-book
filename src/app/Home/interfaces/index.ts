export interface AppHomeAction {
  type: string;
  payload?: object;
}