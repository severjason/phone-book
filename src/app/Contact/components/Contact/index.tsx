import * as React from 'react';
import { AppContact } from '../../interfaces';
import { AppSearchDispatch } from '../../../Nav/interfaces';
import { Card, CardHeader, CardContent, Typography, Divider, CardActionArea } from '@material-ui/core';
import { Phone } from '@material-ui/icons';
import ContactStyles from './styles';
import { ContactButtons } from '../../components';

class Contact extends React.PureComponent<AppContact & AppSearchDispatch, {}> {

  private getHeader(): React.ReactNode {
    const {name, id, toggleContact} = this.props;
    return (
      <CardHeader
        className={`contact-header`}
        title={`${name.first}  ${name.last}`}
        onClick={() => toggleContact(id)}
      />
    );
  }

  public render() {
    const {phone, id, deleteContact, expanded} = this.props;
    return (
      <ContactStyles>
        <Card className="contact-card">
          <CardActionArea className={`card-action`}>
            {this.getHeader()}
          </CardActionArea>
          <Divider/>
          {expanded && <CardContent className={`contact-content expanded`}>
            <Typography component="div">
              <Typography component="div" className={'phones-header'}>
                {phone.length > 1 ? 'Phones:' : 'Phone:'}
              </Typography>
              <div className={`phones-container`}>
                <Phone/>{phone.map((phone: any) => <div key={phone}>{phone}</div>)}
              </div>
            </Typography>
            <Divider/>
            <div className="buttons-container">
              <ContactButtons contactId={id} deleteContact={deleteContact}/>
            </div>
          </CardContent>}
        </Card>
      </ContactStyles>
    );
  }
}

export default Contact;
