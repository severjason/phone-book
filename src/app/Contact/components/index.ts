import Contact from './Contact';
import ContactButtons from './ContactButtons';
import AddContactForm from './AddContact';

export {
  Contact,
  ContactButtons,
  AddContactForm,
};
