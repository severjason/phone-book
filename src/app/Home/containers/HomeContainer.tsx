import * as React from 'react';
import { connect } from 'react-redux';
import { fetchContacts } from '../redux/actions';
import { AppHomeDispatch, AppHomeProps } from '../interfaces';
import { AppState } from '../../../store/interfaces';
import { getSortedContactsWithDividers } from '../redux/selectors';
import { Home } from '../components';
import { ErrorPage } from '../../common';

class HomeContainer extends React.Component<AppHomeProps & AppHomeDispatch, {}> {

  public componentDidMount() {
    const {fetchContacts, contacts} = this.props;
    if (contacts.length === 0) {
      fetchContacts();
    }
  }

  public render() {
    const {contacts, isLoading, error} = this.props;
    return (
      error ? <ErrorPage error={error}/> : <Home contacts={contacts} isLoading={isLoading}/>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  isLoading: state.contactsState.isLoading,
  contacts: getSortedContactsWithDividers(state.contactsState),
  error: state.contactsState.error,
});

export default connect<AppHomeProps, AppHomeDispatch>(
  mapStateToProps,
  {
    fetchContacts,
  })(HomeContainer);
