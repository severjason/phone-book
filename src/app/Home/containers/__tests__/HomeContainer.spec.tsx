import * as React from 'react';
import { HomeContainer } from '../HomeContainer';
import { shallow } from 'enzyme';
import { AppContact } from '../../../Contact/interfaces';
import { AppHomeDispatch, AppHomeProps } from '../../interfaces';

describe('Home container test', () => {
  const mockContacts: AppContact[] = [
    {
      id: 1,
      name: {
        first: 'man',
        last: 'awesome',
      },
      phone: [],
      expanded: false,
    },
    {
      id: 2,
      name: {
        first: 'man',
        last: 'awesome',
      },
      phone: [],
      expanded: false,
    }
  ];

  it('ConnectedHomeContainer should renders without crashing with initial state', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: [],
      error: null,
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    const instance = wrapper.instance();
    expect(wrapper).toMatchSnapshot();
    // @ts-ignore
    expect(instance.props.fetchContacts).toBeCalled();
  });

  it('ConnectedHomeContainer should renders without crashing with contacts', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: mockContacts,
      error: null,
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    const instance = wrapper.instance();
    expect(wrapper).toMatchSnapshot();
    // @ts-ignore
    expect(instance.props.fetchContacts).not.toBeCalled();
  });

  it('ConnectedHomeContainer should renders with error', () => {
    const mockProps: AppHomeProps & AppHomeDispatch = {
      isLoading: false,
      contacts: [],
      error: 'error',
      fetchContacts: jest.fn(),
      toggleContact: jest.fn(),
      deleteContact: jest.fn(),
    };
    const wrapper = shallow(<HomeContainer {...mockProps}/>);
    expect(wrapper).toMatchSnapshot();
  });
});