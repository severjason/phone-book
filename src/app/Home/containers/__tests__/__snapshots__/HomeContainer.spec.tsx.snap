// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container test ConnectedHomeContainer should renders with error 1`] = `
<ErrorPage
  error="error"
/>
`;

exports[`Home container test ConnectedHomeContainer should renders without crashing with contacts 1`] = `
<Home
  contacts={
    Array [
      Object {
        "expanded": false,
        "id": 1,
        "name": Object {
          "first": "man",
          "last": "awesome",
        },
        "phone": Array [],
      },
      Object {
        "expanded": false,
        "id": 2,
        "name": Object {
          "first": "man",
          "last": "awesome",
        },
        "phone": Array [],
      },
    ]
  }
  deleteContact={[MockFunction]}
  isLoading={false}
  toggleContact={[MockFunction]}
/>
`;

exports[`Home container test ConnectedHomeContainer should renders without crashing with initial state 1`] = `
<Home
  contacts={Array []}
  deleteContact={[MockFunction]}
  isLoading={false}
  toggleContact={[MockFunction]}
/>
`;
